Mathlib/Combinatorics/Nullstellensatz.lean

```diff
@@ -0,0 +1,312 @@
+    {w : τ → ℕ} {P : MvPolynomial σ R} (he : Function.Injective e) :
+    weightedTotalDegree w ((rename e) P) = weightedTotalDegree (w ∘ e) P := by
+  unfold weightedTotalDegree
+  rw [support_rename_of_injective he, Finset.sup_image]
+  congr; ext; unfold weight; simp
+
+theorem eq_zero_of_eval_zero_at_prod {σ : Type*} [Finite σ] [IsDomain R]
+    (P : MvPolynomial σ R) (S : σ → Set R)
+    (Hdeg : ∀ i, P.degreeOf i < (S i).ncard)
+    (Heval : ∀ (x : σ → R), (∀ i, x i ∈ S i) → eval x P = 0) :
+    P = 0 := by
+  obtain ⟨n, ⟨e⟩⟩ := Finite.exists_equiv_fin σ
+  suffices MvPolynomial.rename e P = 0 by
+    have that := MvPolynomial.rename_injective (R := R) e (e.injective)
+    rw [RingHom.injective_iff_ker_eq_bot] at that
+    rwa [← RingHom.mem_ker, that] at this
+  apply eq_zero_of_eval_zero_at_prod_nat _ (fun i ↦ S (e.symm i))
+  · intro i
+    classical
+    convert Hdeg (e.symm i)
+    conv_lhs => rw [← e.apply_symm_apply i, degreeOf_rename_of_injective e.injective]
+  · intro x hx
+    simp only [MvPolynomial.eval_rename]
+    apply Heval
+    intro s
+    simp only [Function.comp_apply]
+    convert hx (e s)
+    simp only [Equiv.symm_apply_apply]
+
+open MonomialOrder
+
+lemma _root_.MonomialOrder.degree_binomial [Nontrivial R]
+    {ι : Type*} (m : MonomialOrder ι) (i : ι) (r : R) :
+    m.degree (X i - C r) = single i 1 := by
+  rw [degree_sub_of_lt, degree_X]
+  simp only [degree_C, map_zero, degree_X]
+  rw [← bot_eq_zero, bot_lt_iff_ne_bot, bot_eq_zero, ← map_zero m.toSyn]
+  simp
+
+lemma _root_.MonomialOrder.leadingCoeff_binomial {ι : Type*} (m : MonomialOrder ι) (i : ι) (r : R) :
+    m.leadingCoeff (X i - C r) = 1 := by
+  classical
+  by_cases H : Nontrivial R
+  · simp only [leadingCoeff, m.degree_binomial i r]
+    simp only [coeff_sub, coeff_single_X, true_and, if_true, coeff_C, sub_eq_self]
+    rw [if_neg]
+    intro h
+    apply zero_ne_one (α := ℕ)
+    simp only [Finsupp.ext_iff, coe_zero, Pi.zero_apply] at h
+    rw [h i, single_eq_same]
+  · by_contra H'
+    exact H (nontrivial_of_ne _ _ H')
+
+theorem _root_.MonomialOrder.prod_degree [Nontrivial R]
+    {ι : Type*} (m : MonomialOrder ι) (i : ι) (s : Finset R) :
+    m.degree (s.prod (fun r ↦ X i - C r)) = single i s.card := by
+  classical
+  have H : ∀ r ∈ s, m.degree (X i - C r) = single i 1 := by
+    intro r _
+    rw [degree_sub_of_lt, degree_X]
+    simp only [degree_C, map_zero, degree_X]
+    rw [← bot_eq_zero, bot_lt_iff_ne_bot, bot_eq_zero, ← map_zero m.toSyn]
+    simp
+  rw [MonomialOrder.degree_prod_of_regular]
+  · rw [Finset.sum_congr rfl H]
+    simp only [Finset.sum_const, smul_single, smul_eq_mul, mul_one]
+  · intro r hr
+    convert isRegular_one
+    simp only [leadingCoeff, H r hr]
+    simp only [coeff_sub, coeff_single_X, true_and, if_true, coeff_C, sub_eq_self]
+    rw [if_neg]
+    intro h
+    apply zero_ne_one (α := ℕ)
+    simp only [Finsupp.ext_iff, coe_zero, Pi.zero_apply] at h
+    rw [h i, single_eq_same]
```

Comment:
Can these move earlier?

Reply:
I moved the first two to MonomialOrder and deleted the last one (it isn't used)